src/GafferScene/Instancer.cpp

```diff
@@ -323,7 +337,103 @@ class Instancer::EngineData : public Data
 			}
 		}
 
+		void initPrototypes( PrototypeMode &mode, const std::string &index, const std::string &rootsVariable, const StringVectorData *rootsList )
+		{
+			const std::vector<std::string> *rootStrings = nullptr;
+
+			switch( mode )
+			{
+				case IndexedRootsList :
+				{
+					if( const auto *indices = m_primitive->variableData<IntVectorData>( index ) )
+					{
+						m_indices = &indices->readable();
+						if( m_indices->size() != numPoints() )
+						{
+							throw IECore::Exception( "prototypeIndex primitive variable has incorrect size" );
+						}
+					}
+
+					rootStrings = &rootsList->readable();
+
+					break;
+				}
+				case IndexedRootsVariable :
+				{
+					if( const auto *indices = m_primitive->variableData<IntVectorData>( index ) )
+					{
+						m_indices = &indices->readable();
+						if( m_indices->size() != numPoints() )
+						{
+							throw IECore::Exception( "prototypeIndex primitive variable has incorrect size" );
+						}
+					}
+
+					const auto *roots = m_primitive->variableData<StringVectorData>( rootsVariable, PrimitiveVariable::Constant );
+					if( !roots )
+					{
+						throw IECore::Exception( "prototypeRoots primitive variable must be Constant StringVectorData when using IndexedRootsVariable mode" );
```

Comment:
Including the name makes sense for sure, but I'm not sure its worth distinguishing doesn't exist from wrong type. We've had feedback that its better to give the information up front. Otherwise someone will go create the missing variable only to realize they've made the wrong type or the wrong interpolation. I suppose I could do an additional check after we know we're throwing just to extend the message a bit to say it whether exists or not. Sound ok?

Reply:
Sounds great.

Reply:
Addressed in d08e626df406054344b7c49c745f32c52c5479fc